---
title: "Part 2: Hardy-Weinberg equilibrium"
author: "Arslan A. Zaidi"
date: "2023-07-17"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
```

# Introduction

The purpose of this guide is to illustrate the Hardy-Wenberg principle through simulations in R. The guide assumes that you have a basic knowledge of genetics and have completed Part 1 of the module.

# Some basic terminology 

*Locus:* A position in the genome.

*Allele:* A different variant at a locus. For example, you might carry the 'A' allele at a locus and someone else might carry the 'G' allele.

*Ploidy:* The number of copies of a genome inside the cell of an organism. For example, we are diploid (i.e, have two copies of our genome)

*Genotype:* The genetic makeup of an individual (e.g. 'AA', 'AG', or 'GG' for a diploid organism)

*Genotype frequency:* The fraction of individuals in the population with a particular genotype. For example, we refer to the proportion of individuals with the 'AG' genotype as $f_{AG} = \frac{N_{AG}}{N}$ where $N_{AG}$ is the number of individuals with the 'AG' genotype and $N$ is the total number of individuals in the population.

*Allele frequency:* The no. of copies of an allele in the population. For example, the frequency of the 'A' allele is $f_{A}$. We can calculate this from the genotype frequency like so:

$$ f_{A} = \frac{N_{Aa} + 2N_{AA}}{2N} $$
**Q.** Can you figure out how we came up with this formula?

**Q.** Can we do the reverse? Can we figure out the genotype frequency from the allele frequency?

# Using allele frequency to predict genotype frequency

Let's test this out by simulating some genotypes in a population where the allele frequency is known. Assume the organism is diploid. Then, we need to do the following:

1. Draw one allele at random from the population with some probability. This probability is equal to the frequency of that allele in the population.
2. Draw a second allele from the population with the same probability. 
3. Combine the two alleles to create a genotype.

Let's try this. Assume the frequency of the A allele is 0.5. What's the frequency of the G allele?

```{r}

ninds = 10
fa = 0.5
fg = 1-fa
# Since we are diploid, we will draw one allele at a time
# we code the genotypes numerically for convenience: A -> 1 and G -> 0
# This coding indicates the presence/absence of the A allele

# the sample function draws at random from the different choices given
# based on some probability
# Draw the first allele
x1 = sample(c(0,1), 
            size = ninds, 
            replace = TRUE, 
            prob = c(fa, fg))

#second allele
x2 = sample(c(0,1), 
            size = ninds, 
            replace = TRUE, 
            prob = c(fa, fg))

# Dipolid genotype = no. of copies (out of 2) of the A allele
x = x1 + x2

```

**Q.** Use your R skills to calculate the proportion of each genotype. Hint: you can use the "which()" and "length()" functions.

**Q.** Is there a pattern? Why do you think that is? Does the result change if you rerun the last block of code? Does it change a lot or little?

Try rerunning the last block of code a number of times with a larger sample size (N = 1,000).

```{r}

# code here

```


**Q.** Is the pattern more apparent? Are the values more consistent? Why do you think that is? Do they converge to some particular set of values?

Let's do this more systematically by replicating the result multiple times and averaging over replicates. Let's also try a different starting allele frequency.

```{r}

# Let's run it a couple of times and store the result for each time

# we can create less work for ourselves by writing 'functions'
# functions are a convenient way to run the same thing multiple times
# a function requires an 'argument' - something that we specify
# in this case, the number of individuals and the allele frequency
create_genotype = function(ninds, fa){

  fg = 1-fa
  x1 = sample(c(0,1), 
              size = ninds, 
              replace = TRUE, 
              prob = c(fa, fg))
  
  x2 = sample(c(0,1), 
              size = ninds, 
              replace = TRUE, 
              prob = c(fa, fg))
  
  x = x1 + x2
  # provide an output using return()
  #prop. of GG genotypes
  n.gg = mean(x == 0) # another way to calculate proportions
  n.ga = mean(x == 1)
  n.aa = mean(x == 2)
  return(c(n.gg, n.ga, n.aa))
}

# create an empty matrix to store results in
# 1000 rows for replicates
# 3 columns, one for each genotype
mat = matrix(NA, nrow = 1000, ncol = 3)
for(i in 1:1000){
  #use the function to run the block of code
  mat[i,] = create_genotype(fa = 0.1, ninds = 1000)
}

#calculate the mean genotype across replicates
apply(mat, 2, mean)

```


**Q.** Is there a pattern? Can you guess how these numbers might be related to the allele frequency? Don't move on if you haven't figured this out yet.

Let's try calculating the genotype frequencies for different allele frequencies and plotting the result.

```{r}

#setup a vector with allele frequencies ranging from 0 to 1
f = seq(0, 1, 0.1)
mat = matrix(NA, nrow = length(f), ncol = 3)
for(i in 1:length(f)){
    mat[i,] = create_genotype(fa = f[i], ninds = 1000)
}

#convert to dataframe
mat = data.frame(mat)
colnames(mat) = c("GG","GA","AA")
mat$freq = f

#plot
ggplot(mat)+
  geom_line(aes(freq, GG), color = "blue")+
  geom_line(aes(freq, GA), color = "green")+
  geom_line(aes(freq, AA), color = "red")+
  theme_classic()+
  labs(x = "Frequency of A allele", 
       y = "Genotype frequency")

```

**Q.** Take a look at this plot and describe what's going on.

That you can predict the genotype frequencies in a population from the allele frequencies is part of the Hardy-Weinberg equilibrium (HWE). Bet there's a blast from the past you never thought you'd be hearing again. This principle states that but if:

1. The population is mating randomly
2. The population size is large 
3. There is no mutation
4. There is no natural selection

Then, the frequency of the GG, AG, and AA genotypes are expected to be equal to: $f_{GG} = f_G^2$, $f_{GA} = 2f_Gf_A$, and $f_{AA} = f_A^2$. Since $f_A = 1 - f_G$, we can simplify this to: 

$f_{GG} = (1 - f_A)^2$, $f_{GA} = 2f_A(1-f_G)$, and $f_{AA} = f_A^2$

That's exactly what the lines in the previous plot show. Let's test this out by adding dotted lines to the previous plot 


```{r}

#the alpha aesthetic makes the lines more transparent so we can see the dotted lines
ggplot(mat)+
  geom_line(aes(freq, GG), color = "blue", alpha = 0.1)+
  geom_line(aes(freq, GA), color = "green", alpha = 0.1)+
  geom_line(aes(freq, AA), color = "red", alpha = 0.1)+
  geom_line(aes(freq, freq^2), 
            color = "blue", linetype = "dotted")+
  geom_line(aes(freq, 2*freq*(1-freq)), 
            color = "green", linetype = "dotted")+
  geom_line(aes(freq, (1-freq)^2), 
            color = "red", linetype = "dotted")+
  theme_classic()+
  labs(x = "Frequency of A allele", 
       y = "Genotype frequency")

#notice how we have layered multiple plots on top of each other.

```

**Q.** Do the observed genotype frequencies overlap with those predicted by HWE? 

Try rerunning the last TWO blocks of code but with 10 individuals instead of 1,000.

```{r}

#code here


#setup a vector with allele frequencies ranging from 0 to 1
f = seq(0, 1, 0.1)
mat = matrix(NA, nrow = length(f), ncol = 3)
for(i in 1:length(f)){
    mat[i,] = create_genotype(fa = f[i], ninds = 10)
}

#convert to dataframe
mat = data.frame(mat)
colnames(mat) = c("GG","GA","AA")
mat$freq = f

ggplot(mat)+
  geom_line(aes(freq, GG), color = "blue", alpha = 0.1)+
  geom_line(aes(freq, GA), color = "green", alpha = 0.1)+
  geom_line(aes(freq, AA), color = "red", alpha = 0.1)+
  geom_line(aes(freq, freq^2), 
            color = "blue", linetype = "dotted")+
  geom_line(aes(freq, 2*freq*(1-freq)), 
            color = "green", linetype = "dotted")+
  geom_line(aes(freq, (1-freq)^2), 
            color = "red", linetype = "dotted")+
  theme_classic()+
  labs(x = "Frequency of A allele", 
       y = "Genotype frequency")


```

**Q.** What do you observe? Why do you think this is? Hint: this is related to one of the assumptions of HWE.

You see, we can use mathematics to approximate what goes on in populations just like we did by approximating the expected genotype frequencies in the population from allele frequencies. Obviously, the assumptions are important and can sometimes breakdown. But first we must understand highly idealized situations as a place to start.

The Hardy-Weinberg equilibrium does not only give us a way to predict genotype frequencies from allele frequencies but also predicts that allele frequencies will NOT change in populations over time if the four conditions are met. This again provides us a null model to study evolutionary processes, which are all about change!

The function we wrote to simulate genotypes based on allele frequencies is kind of clunky and R has built-in functions to do this: rbinom(). To simulate the (diploid) genotypes of 100 individuals from a population where the allele frequency is 0.15, we do the following:

```{r}


#rbinom takes 3 arguments
#n = no. of individuals
#size = ploidy
#prob = allele frequency
x = rbinom(100, 2, 0.15)

#output the frequency of each genotype
mean(x == 0)
mean(x == 1)
mean(x == 2)

#OR we could use the table() function
table(x)/100

```


Congrats, you're done with Part 2! You can move on to the 3rd part. As before, save your work, close "PopGen.Rmd", go to "Session > Clear Workspace", and then "Session > Restart R"








